File: src/day02.js
fs = require("fs")

// const debug = console.log
const debug = () => {}

const isReportSafe = (report, skipIndex = null) => {
    let prev = null
    let expectingIncreasing = null
    const count = report.length
    for (let i = 0; i < count; ++i) {
        const curr = report[i]

        if (skipIndex == i) {
            debug(report, `dampening(${skipIndex}) => ${curr}`)
            continue
        }

        if (prev != null) {
            if (expectingIncreasing == null) {
                expectingIncreasing = prev < curr
                debug(report, `expectingIncreasing: ${expectingIncreasing}`)
            } else {
                isIncreasing = prev < curr
                if (isIncreasing != expectingIncreasing) {
                    debug(report, `Unsafe: isIncreasing(${isIncreasing}) != expectingIncreasing(${expectingIncreasing})`)
                    return false
                }
            }

            const diff = Math.abs(prev - curr)
            if (diff < 1 || diff > 3) {
                debug(report, `Unsafe: diff(${diff})`)
                return false
            }
        }

        prev = curr
    }

    if (skipIndex == null) {
        debug(report, `Safe`)
    } else {
        debug(report, `Safe (dampened)`)
    }
    return true
}

const isDampenedReportSafe = (report) => {
    debug(report, "Dampening...")

    const count = report.length
    for (let i = 0; i < count; ++i) {
        if (isReportSafe(report, i)) {
            return true
        }
    }

    debug(report, `Unsafe, even dampened`)
    return false
}

const main = async (argv) => {
    const inputFilename = argv.shift()
    if (!inputFilename) {
        console.error("Syntax: day02 [input.txt]")
        return
    }

    const lines = fs
        .readFileSync(inputFilename, "utf8")
        .split(/[\r\n]/)
        .filter((line) => line.length > 0)

    let safeCount = 0
    let dampenedSafeCount = 0
    let totalCount = 0
    for (let line of lines) {
        debug(`-- ${line} --`)
        if (line.length < 1) {
            continue
        }

        const report = line.split(/\s+/).map((e) => parseInt(e))
        if (isReportSafe(report)) {
            ++safeCount
        } else if (isDampenedReportSafe(report)) {
            ++dampenedSafeCount
        }
        ++totalCount
    }

    console.log(`Safe (basic)   : ${safeCount}`)
    console.log(`Safe (dampened): ${dampenedSafeCount}`)
    console.log(`Safe (combined): ${safeCount + dampenedSafeCount}`)
    console.log(`Total          : ${totalCount}`)
}

main(process.argv.slice(2))

File: src/day02.rs
fn parse_report(line: &str) -> Vec<i32> {
    let nums = line.split_whitespace();
    let mut report = Vec::<i32>::new();

    for num_string in nums {
        let num = num_string.parse::<i32>().unwrap();
        report.push(num);
    }

    return report;
}

fn is_report_safe(report: &Vec<i32>, skip_index: Option<usize>) -> bool {
    let count = report.len();
    let mut detecting_increasing = true;
    let mut expecting_increasing = false;
    let mut prev = report[0];
    let mut i = 1;

    if skip_index.is_some() {
        if skip_index.unwrap() == 0 {
            prev = report[1];
            i += 1;
        }
    }

    while i < count {
        let curr = report[i];

        if skip_index.is_some() {
            if skip_index.unwrap() == i {
                i += 1;
                continue;
            }
        }

        if detecting_increasing {
            detecting_increasing = false;
            expecting_increasing = prev < curr;
        } else {
            let is_increasing = prev < curr;
            if is_increasing != expecting_increasing {
                return false;
            }
        }

        let diff = (prev - curr).abs();
        if diff < 1 || diff > 3 {
            return false;
        }

        prev = curr;
        i += 1;
    }

    return true;
}

fn is_dampened_report_safe(report: &Vec<i32>) -> bool {
    let count = report.len();
    for i in 0..count {
        if is_report_safe(report, Some(i)) {
            return true;
        }
    }

    return false;
}

pub fn main(contents: String) {
    let mut safe_count = 0;
    let mut dampened_count = 0;
    let mut total_count = 0;

    for line in contents.lines() {
        let report = parse_report(line);
        if report.len() == 0 {
            continue;
        }

        if is_report_safe(&report, None) {
            safe_count += 1;
        } else if is_dampened_report_safe(&report) {
            dampened_count += 1;
        }

        total_count += 1;
    }

    println!("Safe (basic)   : {}", safe_count);
    println!("Safe (dampened): {}", dampened_count);
    println!("Safe (combined): {}", safe_count + dampened_count);
    println!("Total          : {}", total_count);
}
